splunk_otel/profiling/__init__.py

```diff
@@ -0,0 +1,300 @@
+            fun.name = name_id
+            fun.system_name = name_id
+            fun.filename = str_table.index(file_name)
+            functions_table[key] = fun
+
+        return fun
+
+    def get_line(file_name, function_name, line_no):
+        line = profile_pb2.Line()
+        line.function_id = get_function(file_name, function_name).id
+        line.line = line_no if line_no != 0 else -1
+        return line
+
+    def get_location(frame):
+        (file_name, function_name, line_no) = frame
+        key = f"{file_name}:{function_name}:{line_no}"
+        location = locations_table.get(key)
+
+        if location is None:
+            location = profile_pb2.Location()
+            location.id = len(locations_table)
+            location.line.append(get_line(file_name, function_name, line_no))
+            locations_table[key] = location
+
+        return location
+
+    timestamp_key = str_table.index("source.event.time")
+    trace_id_key = str_table.index("trace_id")
+    span_id_key = str_table.index("span_id")
+    thread_id_key = str_table.index("thread.id")
+    event_period_key = str_table.index("source.event.period")
+
+    pb_profile = profile_pb2.Profile()
+
+    event_period_label = profile_pb2.Label()
+    event_period_label.key = event_period_key
+    event_period_label.num = interval
+
+    samples = []
+    for stacktrace in stacktraces:
+        thread_id = stacktrace["tid"]
+
+        timestamp_label = profile_pb2.Label()
+        timestamp_label.key = timestamp_key
+        timestamp_label.num = int(stacktrace["timestamp"] / 1e6)
+
+        thread_id_label = profile_pb2.Label()
+        thread_id_label.key = thread_id_key
+        thread_id_label.num = thread_id
+
+        labels = [timestamp_label, event_period_label, thread_id_label]
+
+        trace_context = profiler.thread_states.get(thread_id)
+        if trace_context:
+            (trace_id, span_id) = trace_context
+
+            trace_id_label = profile_pb2.Label()
+            trace_id_label.key = trace_id_key
+            trace_id_label.str = str_table.index(f"{trace_id:#016x}")
+            labels.append(trace_id_label)
+
+            span_id_label = profile_pb2.Label()
+            span_id_label.key = span_id_key
+            span_id_label.str = str_table.index(f"{span_id:#08x}")
+            labels.append(span_id_label)
+
+        sample = profile_pb2.Sample()
+
+        location_ids = []
+
+        for frame in reversed(stacktrace["stacktrace"]):
+            location_ids.append(get_location(frame).id)
+
+        sample.location_id.extend(location_ids)
+        sample.label.extend(labels)
+
+        samples.append(sample)
+
+    pb_profile.sample.extend(samples)
+    pb_profile.string_table.extend(str_table.keys())
+    pb_profile.function.extend(list(functions_table.values()))
+    pb_profile.location.extend(list(locations_table.values()))
+
+    return gzip.compress(pb_profile.SerializeToString())
+
+
+def _profiler_loop(options: _Options):
```

Comment:
What do you think about making this functionality unit-testable?

Reply:
Split this function into multiple smaller ones now, I think it is currently implicitly covered by our unit tests